Firestore/core/src/model/object_value.cc

```diff
@@ -0,0 +1,287 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/model/object_value.h"
+
+#include <map>
+#include <set>
+
+#include "Firestore/Protos/nanopb/google/firestore/v1/document.nanopb.h"
+#include "Firestore/core/src/nanopb/message.h"
+#include "Firestore/core/src/nanopb/nanopb_util.h"
+
+namespace firebase {
+namespace firestore {
+namespace model {
+
+MutableObjectValue::MutableObjectValue() {
+  value_->which_value_type = google_firestore_v1_Value_map_value_tag;
+  value_->map_value.fields_count = 0;
+  value_->map_value.fields =
+      nanopb::MakeArray<google_firestore_v1_MapValue_FieldsEntry>(0);
+}
+
+model::FieldMask MutableObjectValue::ToFieldMask() const {
+  return ExtractFieldMask(value_->map_value);
+}
+
+model::FieldMask MutableObjectValue::ExtractFieldMask(
+    const google_firestore_v1_MapValue& value) const {
+  std::set<FieldPath> fields;
+  for (size_t i = 0; i < value.fields_count; ++i) {
+    FieldPath current_path({nanopb::MakeString(value.fields[i].key)});
+    if (value.fields[i].value.which_value_type ==
+        google_firestore_v1_Value_map_value_tag) {
+      // Recursively extract the nested map
+      model::FieldMask nested_mask =
+          ExtractFieldMask(value.fields[i].value.map_value);
+      if (nested_mask.begin() == nested_mask.end()) {
+        // Preserve the empty map by adding it to the FieldMask.
+        fields.insert(current_path);
+      } else {
+        for (const FieldPath& nested_path : nested_mask) {
+          fields.insert(current_path.Append(nested_path));
+        }
+      }
+    } else {
+      fields.insert(current_path);
+    }
+  }
+  return model::FieldMask(fields);
+}
+
+absl::optional<google_firestore_v1_Value> MutableObjectValue::Get(
+    const firebase::firestore::model::FieldPath& path) const {
+  if (path.empty()) {
+    return *value_;
+  } else {
+    google_firestore_v1_Value nested_value = *value_;
+    for (const std::string& segment : path) {
+      _google_firestore_v1_MapValue_FieldsEntry* entry =
+          FindEntry(nested_value, segment);
+      if (!entry) return {};
+      nested_value = entry->value;
+    }
+    return nested_value;
+  }
+}
+
+void MutableObjectValue::Set(const model::FieldPath& path,
+                             const google_firestore_v1_Value& value) {
+  HARD_ASSERT(!path.empty(), "Cannot set field for empty path on ObjectValue");
+
+  _google_firestore_v1_MapValue* parent_map = ParentMap(path.PopLast());
+
+  std::string last_segment = path.last_segment();
+  std::map<std::string, google_firestore_v1_Value> upserts{
+      {last_segment, value}};
+
+  ApplyChanges(parent_map, upserts, /* deletes= */ {});
+}
+
+void MutableObjectValue::SetAll(const model::FieldMask& field_mask,
+                                const MutableObjectValue& data) {
+  FieldPath parent;
+
+  std::map<std::string, google_firestore_v1_Value> upserts;
```

Comment:
I think you can get away with storing pointers to protos in this map -- they would point into `data` which should stay valid for the duration of this function.